OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,293 @@
+/* -------------------------------------------------------------------------- *
+ *                            OpenSim:  DataTable.h                           *
+ * -------------------------------------------------------------------------- *
+ * The OpenSim API is a toolkit for musculoskeletal modeling and simulation.  *
+ * See http://opensim.stanford.edu and the NOTICE file for more information.  *
+ * OpenSim is developed at Stanford University and supported by the US        *
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2015 Stanford University and the Authors                *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#ifndef OPENSIM_DATATABLE_H_
+#define OPENSIM_DATATABLE_H_
+
+/** \file
+This file defines the  DataTable_ class, which is used by OpenSim to provide an 
+in-memory container for data access and manipulation.                         */
+
+// Non-standard headers.
+#include "SimTKcommon.h"
+#include "OpenSim/Common/Exception.h"
+#include "OpenSim/Common/ValueArrayDictionary.h"
+
+namespace OpenSim {
+
+/** AbstractDataTable is the base-class of all DataTable_(templated) allowing 
+storage of DataTable_ templated on different types to be stored in a container 
+like std::vector. AbstractDataTable_ offers:
+- Interface to access column metadata of the DataTable_.
+- A heterogeneous container to store metadata associated with the DataTable_ in
+  the form of key-value pairs where key is of type std::string and value can be
+  of any type.
+
```

Comment:
Nit picky, perhaps, but I generally do not like the description of a base class to describe derived classes other than as part of an example. AbstractDataTable defines the interface for all data tables and should articulate that interface clearly. For example, it defines data table meta data including meta data for each column in the table. It also defines the table structure consisting of a uniform column of independent data and homogeneous columns of dependent data. I think these are important aspects to describe in more detail.